codec/helper.go

```diff
@@ -1064,6 +1122,9 @@ type typeInfo struct {
 	cs  bool // T is a Selfer
 	csp bool // *T is a Selfer
 
+	ufh  bool // T is an UnknownFieldHandler
```

Comment:
Everything above this point in the file was copied over with no changes; everything here and below had to be reworked.